doc/iscsi-iname.8
.TH ISCSI_INAME 8 "Jan 2010" "" "Linux Administrator's Manual"
.SH NAME
iscsi-iname \- iSCSI initiator name generation tool
.SH SYNOPSIS
.BI iscsi-iname
[OPTION]
.SH "DESCRIPTION"
.B iscsi-iname
generates a unique iSCSI node name on every invocation.

.SH OPTIONS
.TP
.BI [-h|--help]
Display help
.TP
.BI [-p=]\fIprefix\fP
Use the prefix passed in instead of the default "iqn.2016-04.com.open-iscsi"

.SH AUTHORS
Open-iSCSI project <http://www.open-iscsi.com/>
.br
Mike Christie <mchristi@redhat.com>


Changelog
open-iscsi-2.0-877 - open-iscsi-2.0.878

# output from "git shortlog --no-merges 2.0.877..HEAD"

Daniel Schaefer (1):
      Add target to install systemd units

Lee Duncan (18):
      iscsistart is not installed
      Fix i586 build issues with string length overflow.
      Use pkg-config in Makefiles for newer libraries.
      Updated iscsiadm man page: add "onboot" handling.
      Fix output for iscsiadm node/iface print level P1
      When displaying interfaces, skip "iface.example"
      Fix node print return value when no nodes.
      Fix printing of node database again.
      Fix output of node printing for multiple paths.
      Stop using /var directory for PIDfile and locks
      Improve daemon synchronization, fix err msgs
      Fix pipe notification code
      Add systemd support for iscsiuio
      Make iscsid systemd usage optional
      Fix possible discovery hang when timing out
      Handle systemd disablement correctly in iscsiuio
      The iscsi login/logout service requires iscsid.
      Remove redundant Requires= from iscsi.service

Manish Rangankar (3):
      iscsiuio: Do not flush tx queue on each uio interrupt.
      qedi: Set buf_size in case of ICMP and ARP packet.
      qedi: Use uio BD index instead on buffer index.

Nilesh Javali (3):
      iscsiuio: v0.7.8.5
      iscsiuio: allow processing of iscsid requests in DHCP failure condition
      iscsiuio: update version to 0.7.8.6

Xiubo Li (1):
      rec update: disable the idbm_lock in read/write when updating the rec

fredvx (1):
      Add Restart=on-failure option to iscsid.service

igo95862 (2):
      Make iscsid.service a requirement.
      Fixed iscsi.service considering every signal and exit code as successful. Now only code 21 (no objects found to execute on) and normal exit conditions are valid.

open-iscsi-2.0-876 - open-iscsi-2.0.877

Antoine de Maleprade (1):
      iscsid: fix logging level when starting daemon

Cathy Zhou (1):
      Reduce delays to improve iscsi boot performance

Chris Leech (16):
      libopeniscsiusr: fixes err on prefix_len
      vlan setting sync across ipv4/ipv6 for be2iscsi
      iscsid logging blank messages at level EMERG
      iscsistart: prevent unix socket cross-talk
      libopeniscsiusr: hosts can have multiple ifaces
      libopeniscsiusr: clear errno before calling strtoll
      libopeniscsiusr: setup ipv6 records based on iface name
      libopeniscsiusr: use asprintf and remove PATH_MAX stack buffers
      Merge pull request #110 from cleech/libopeniscsiusr_fixes
      Merge pull request #111 from cleech/for_upstream
      Merge pull request #106 from phmccarty/lib-symlinks
      enable MaxOutstandingR2T negotiation
      Merge pull request #114 from njavali/iscsiuio-bug-fixes
      Merge pull request #115 from cleech/for_upstream
      Merge pull request #120 from maxnet/va
      Merge pull request #122 from njavali/iscsiuio-bug-fixes

Christian Ehrhardt (2):
      iscsiuio: avoid loosing bad rc in nic_nl_open
      iscsiuio: fail on nic_nl_open failing

Floris Bos (1):
      context.h: add missing stdarg.h include

Gris Ge (8):
      libopeniscsiusr: Fix iscsi_sessions_free() on 0 se_count.
      libopeniscsiusr: Add full iscsi interface support.
      Makefile: Trivial change on aligning output.
      libopeniscsiusr: Fix incorrect debug message for iface query.
      libopeniscsiusr: Add node query support
      libopeniscsiusr: Fix iscsi_iface_get() on default interfaces.
      iscsiadm: Use libopeniscsiusr in `iscsiadm -m iface -P1`
      libopeniscsiusr: Fix compile error on GCC 8.

Khem Raj (6):
      libopeniscsiusr: Include limit.h for PATH_MAX
      libopeniscsiusr: Add CFLAGS to linker cmdline
      qedi.c: Removed unused linux/ethtool.h
      idbm.c: Include fcnl.h for O_RDWR and O_CREAT definitions
      bnx2x.c: Reorder the includes to avoid duplicate defines with musl
      fwparam_ppc.c: Do not use __compar_fn_t

Lee Duncan (55):
      Merge pull request #91 from njavali/iscsiuio-bug-fixes
      Merge pull request #92 from az0uz/master
      Use correct size when copying nic name.
      Do not overload global sysfs_path locally.
      libopeniscsiusr: ensure sysfs pathname doesn't overflow.
      Ensure sysfs pathname doesn't overflow.
      Merge pull request #89 from kraj/kraj/musl-fixes
      Merge pull request #94 from gonzoleeman/gcc-8-fixes
      Merge branch 'master' into api
      Merge pull request #93 from cathay4t/api
      Keep iscsi_if in sync with kernel version.
      Add a TODO item on iscsi_if.h.
      Merge pull request #98 from gonzoleeman/update-iscsi_if-to-latest
      Add error message for new ISCSI_ERR_NOP_TIMEDOUT
      Merge pull request #99 from gonzoleeman/handle-new-kernel-error-code
      Allow a host_id value of zero.
      Merge pull request #100 from gonzoleeman/libopeniscsiusr-allow-zero-host_id
      Merge pull request #101 from gonzoleeman/fix-qedi-iface-name
      Fix iscsiuio segfault when shutting down.
      Remove unused file fwparam_ibft_sysfs.c.
      Merge pull request #96 from cathay4t/api
      Merge pull request #103 from cathy-zhou/perf_fix
      Plugging a memory leak from discovery.
      Merge pull request #123 from gonzoleeman/fix-discovery-leak
      Allow reading sysfs "port" to fail gracefully.
      Fix incorrect sysfs logic for port and ip address.
      Fix reading of sysfs signed integers when negative.
      Create a new error for "target not connected".
      Fix bug in error message when reading sysfs numbers.
      Handle ENOTCONN error separately when reading sysfs values.
      Limit session relogin attempts using config value.
      Merge pull request #127 from smoser/fix/iscsid-pidfile-write
      Merge pull request #129 from cpaelzer/cleanup-nic_nl_open-usage
      Merge pull request #131 from gonzoleeman/fix-reconnect-forever
      Removed unused value 'one'.
      Include stdio.h for use of snprintf().
      Fix qsort() comparator function call.

utils/iscsi-iname.c
/*
 * iSCSI InitiatorName creation utility
 * Copyright (C) 2001 Cisco Systems, Inc.
 * maintained by linux-iscsi-devel@lists.sourceforge.net
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * See the file COPYING included with this distribution for more details.
 *
 * $Id: iscsi-iname.c,v 1.1.2.3 2005/03/15 06:33:44 wysochanski Exp $
 *
 * iscsi-iname.c - Compute an iSCSI InitiatorName for this host.
 * Note that to ensure uniqueness, the system time is
 * a factor.  This name must be cached and only regenerated
 * if there is no cached value.
 */

#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/utsname.h>
#include <sys/time.h>

#include "md5.h"

#define RANDOM_NUM_GENERATOR	"/dev/urandom"

/* iSCSI names have a maximum length of 223 characters, we reserve 13 to append
 * a seperator and 12 characters (6 random bytes in hex representation) */
#define PREFIX_MAX_LEN 210

int
main(int argc, char *argv[])
{
	struct timeval time;
	struct utsname system_info;
	long hostid;
	struct MD5Context context;
	unsigned char digest[16];
	unsigned char *bytes = digest;
	unsigned char entropy[16];
	int e;
	int fd;
	char *prefix;

	/* initialize */
	memset(digest, 0, sizeof (digest));
	memset(&context, 0, sizeof (context));
	MD5Init(&context);

	/* take a prefix if given, otherwise use a default. */
	if (argc > 1 && argv[1]) {
		prefix = argv[1];
		if (( strcmp(prefix, "-h") == 0 ) ||
		    ( strcmp(prefix, "--help") == 0 )) {
			printf("\nGenerates a unique iSCSI node name "
			       "on every invocation.\n");
			exit(0);
		} else if ( strcmp(prefix, "-p") == 0 ) {
			prefix = argv[2];
			if (strnlen(prefix, PREFIX_MAX_LEN + 1) > PREFIX_MAX_LEN) {
				printf("Error: Prefix cannot exceed %d "
				       "characters.\n", PREFIX_MAX_LEN);
				exit(1);
			}
		} else {
			printf("\nUsage: iscsi-iname [-h | --help | "
			       "-p <prefix>]\n");
			exit(0);
		}
	} else {
		prefix = "iqn.2016-04.com.open-iscsi";
	}

	/* try to feed some entropy from the pool to MD5 in order to get
	 * uniqueness properties
	 */

	if ((fd = open(RANDOM_NUM_GENERATOR, O_RDONLY))) {
		e = read(fd, &entropy, 16);
		if (e >= 1)
			MD5Update(&context, (md5byte *)entropy, e);
		close(fd);
	}

	/* time the name is created is a factor in order to get
	 * uniqueness properties
	 */
	if (gettimeofday(&time, NULL) < 0) {
		perror("error: gettimeofday failed");
		return 1;
	}
	MD5Update(&context, (md5byte *) & time.tv_sec, sizeof (time.tv_sec));
	MD5Update(&context, (md5byte *) & time.tv_usec, sizeof (time.tv_usec));

	/* hostid */
	hostid = gethostid();
	MD5Update(&context, (md5byte *) & hostid, sizeof (hostid));

	/* get the hostname and system name */
	if (uname(&system_info) < 0) {
		perror("error: uname failed");
		return 1;
	}
	MD5Update(&context, (md5byte *) system_info.sysname,
		  sizeof (system_info.sysname));
	MD5Update(&context, (md5byte *) system_info.nodename,
		  sizeof (system_info.nodename));
	MD5Update(&context, (md5byte *) system_info.release,
		  sizeof (system_info.release));
	MD5Update(&context, (md5byte *) system_info.version,
		  sizeof (system_info.version));
	MD5Update(&context, (md5byte *) system_info.machine,
		  sizeof (system_info.machine));

	/* compute the md5 hash of all the bits we just collected */
	MD5Final(digest, &context);

	/* vary which md5 bytes we pick (though we probably don't need to do
	 * this, since hopefully MD5 produces results such that each byte is as
	 * good as any other).
	 */

	if ((fd = open(RANDOM_NUM_GENERATOR, O_RDONLY))) {
		if (read(fd, entropy, 1) == 1)
			bytes = &digest[(entropy[0] % (sizeof(digest) - 6))];
		close(fd);
	}

	/* print the prefix followed by 6 bytes of the MD5 hash */
	printf("%s:%x%x%x%x%x%x\n", prefix,
		bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5]);
	return 0;
}
